lib/Mir.ml

```diff
@@ -17,6 +17,10 @@ let gensym () =
   _counter := !_counter + 1 ;
   sprintf "sym%d" !_counter
 
+let gensym_enter () =
+  let old_counter = !_counter in
+  (gensym (), fun () -> _counter := old_counter)
+
```

Comment:
I don't think the old code generator generated symbols ?

And yeah, generating fresh symbols should happen centrally somewhere and it's more of a middle/backend thing rather than a front-end thing. The only time I reset the gensym is when I know it's safe to do so, right now that's just when generating for loops for data input checks or similar stuff that doesn't have user code inside it.

Reply:
OK. I think the old code generator just used some naming scheme right? As long as it is used carefully, I'm happy with it.

Reply:
Also, shouldn't the generated names end in `__` or something to make sure they never clash with user variable names?